sunscreen_math/src/metal_impl/ristrettovec.rs

```diff
@@ -0,0 +1,716 @@
+            "ristretto_add",
+            &[&self.data, &rhs.data, &o.data, &len_gpu.data],
+            [(rhs.len() as u64, 64), (1, 1), (1, 1)],
+        );
+
+        o
+    }
+}
+
+impl Sub<RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn sub(self, rhs: RistrettoPointVec) -> Self::Output {
+        &self - &rhs
+    }
+}
+
+impl Sub<&RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn sub(self, rhs: &RistrettoPointVec) -> Self::Output {
+        &self - rhs
+    }
+}
+
+impl Sub<RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn sub(self, rhs: RistrettoPointVec) -> Self::Output {
+        self - &rhs
+    }
+}
+
+impl Sub<&RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn sub(self, rhs: &RistrettoPointVec) -> Self::Output {
+        assert_eq!(self.len(), rhs.len());
+
+        let runtime = Runtime::get();
+
+        let len = self.len_bytes();
+
+        let o = Self::Output {
+            data: runtime.alloc(len),
+            len: rhs.len(),
+        };
+
+        let len_gpu = U32Arg::new(rhs.len() as u32);
+
+        // TODO: o gets mutated here. Need to figure out what that means in terms of UB.
+        runtime.run(
+            "ristretto_sub",
+            &[&self.data, &rhs.data, &o.data, &len_gpu.data],
+            [(rhs.len() as u64, 64), (1, 1), (1, 1)],
+        );
+
+        o
+    }
+}
+
+impl Mul<ScalarVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn mul(self, rhs: ScalarVec) -> Self::Output {
+        &self * &rhs
+    }
+}
+
+impl Mul<&ScalarVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn mul(self, rhs: &ScalarVec) -> Self::Output {
+        &self * rhs
+    }
+}
+
+impl Mul<ScalarVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn mul(self, rhs: ScalarVec) -> Self::Output {
+        self * &rhs
+    }
+}
+
+impl Mul<&ScalarVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn mul(self, rhs: &ScalarVec) -> Self::Output {
+        assert_eq!(self.len(), rhs.len());
+
+        let runtime = Runtime::get();
+
+        let out = RistrettoPointVec {
+            data: runtime.alloc(self.len_bytes()),
+            len: self.len(),
+        };
+
+        let len_gpu = U32Arg::new(self.len() as u32);
+
+        runtime.run(
+            "ristretto_scalar_mul",
+            &[&self.data, &rhs.data, &out.data, &len_gpu.data],
+            [(self.len() as u64, 64), (1, 1), (1, 1)],
+        );
+
+        out
+    }
+}
+
+impl Clone for RistrettoPointVec {
+    fn clone(&self) -> Self {
+        let runtime = Runtime::get();
+
+        let buffer = runtime.alloc(self.len_bytes());
+
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                self.data.contents() as *const u8,
+                buffer.contents() as *mut u8,
+                self.len_bytes(),
+            )
+        };
```

Comment:
Just curious, why do we need this over, say, a default `Clone` impl?

Reply:
I even tested this 😄